contributing: Add contributing to documentation guidelines (#378)
unikraft/docs: content/docs/contributing/docs.mdx

---
title: Contributing to Documentation
---

## Documentation

As with the other parts of Unikraft, we welcome contributions to the [documentation repository](https://github.com/unikraft/docs).

> Documentation is written using [GitHub Markdown](https://github.github.com/gfm/).

Reviewer: Isn't it now [MDX](https://mdxjs.com/)?

We follow the [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow) for contributions.

> Submit corrections and improvements to documentation as pull requests (PRs) that we will then review and approve in the documentation.

Reviewer: ```suggestion
Submit new content, corrections and improvements to documentation as pull requests (PRs) that we will then review and approve in the documentation.
```

You can add entire sections if you think a topic is missing or update existing ones with missing information.

> You can also open up an issue to signal a problem or a missing part in the documentation that you request someone to fix.

Reviewer: ```suggestion
You can also open up [an issue](https://github.com/unikraft/docs/issues) to signal a problem or a missing part in the documentation that you request someone to fix.
```

We use English as the documentation language.
Please proofread your text before submitting a pull request.

Write each sentence on a new line.
This way, changing one sentence only affects one line in the source code.

Prefer using the 1st person plural.
Use phrases like "we run the command / app", "we look at the source code", "we provide support".

### Content Layout

The content has the following structure:

```text
content
|-- blog/                   # Blog posts, <date>-<blogpost title>.md
|-- community/              # Community related content (events, general information, etc.)
|-- docs/                   # Unikraft documentation, split by categories into subdirectories
|-- guides/                 # Short and on point guides, good for getting started
|-- hackathons/             # Hackathon landing pages
|-- imprint.md
|-- privacy.md
`-- releases/               # Unikraft releases summaries

> ```

Reviewer: This needs to be updated to the current layout.

Author: This is the current layout

#### Building the Website

You can build the website inside a Docker environment or natively.
Building inside a Docker environment is **strongly recommended**, since you will work in the same environment as the deployed website.

#### Building the Website Inside a Docker Environment

You can build the site inside a Docker environment.
For this, you will need to install Docker-CE.
You can do that by following the instructions [here](https://github.com/docker/docker-install), or by running:

```console
curl -fsSL https://get.docker.com/ | sh
```

<Info>
Building the site inside a Docker environment is highly recommended, since you will work in the same environment as the deployed site, so there will be no errors regarding packages version or filesystem layout.
</Info>

To deploy the site, run (in the directory of the repository clone):

```console
docker build -t ghcr.io/unikraft/docs:base --target base .

docker run -it --rm -v $(pwd):/docs -w /docs -p 3000:3000 --entrypoint sh ghcr.io/unikraft/docs:base

yarn run dev
```

You may need admin privileges to run the first 2 commands.

> The website will be rendered at `localhost:3000`.

Reviewer: ```suggestion
Open up your browser and point to http://localhost:3000 to check the deployed documentation website.

If you do updates the documentation files (even outside the Docker environment) they will be picked up and rendered at http://localhost:3000.
```

#### Building the Website Natively

You can also build the site natively.
For a native build, you need to install `node`.
You can do it by following the instructions [here](https://nodejs.org/en/download).

<Info>
This may lead to errors due to different `node` versions or other environment problems.
These problems can be avoided by building the website inside a Docker environment.
</Info>

Then, to build and deploy the site, run:

```console
yarn install
yarn run dev
```

> Access the `http://localhost:3000/` URL in the browser to view the site.

Reviewer: ```suggestion
As above, open up your browser and point to http://localhost:3000 to check the deployed documentation website.
```

#### Using the Linter

> To automate a big part of the reviewing process, we use the [`super-linter`](https://github.com/github/super-linter) on the `docs` repository.

Reviewer: ```suggestion
To automate a big part of the reviewing process, we use the [`super-linter`](https://github.com/github/super-linter) on this repository.
```

Author: The content will be viewed on the website, `this` would seem like we are talking about the website, not a repository. I would leave it as it is now, if you want me to change it let me know.

```console
npm install markdownlint-cli
npm install textlint
npm install jscpd
```

Note that it's likely that these linters require a more recent version of `node` than the default package installation.
Install the most recent one using [`nvm`](https://github.com/nvm-sh/nvm).
First, install `nvm` using [the instructions here](https://github.com/nvm-sh/nvm).
Then install the most recent version of `node` and follow the instructions:

```console
nvm install node
# Follow instructions shown by the above command.
nvm use node
node --version # Check node version.
```

The `textlinter` will require installation of rules.
You can find a list of rules [here](https://github.com/textlint/textlint/wiki/Collection-of-textlint-rule).

```console
npm install textlint-rule-terminology
npm install textlint-rule-no-todo
```

Use the linters with `npx` from the documentation directory:

```console
npx markdownlint --config .github/workflows/config/config.json --rules .github/workflows/rules/rules.js path/to/markdown/file.mdx
npx textlint --config .github/workflows/.textlintrc --rule <rule-name> path/to/markdown/file.mdx

> npx jscpd <path> # Tipically a directory

Reviewer: ```suggestion
npx jscpd <path> # Typically a directory
```

---
title: Contributing to Documentation
---

## Documentation

As with the other parts of Unikraft, we welcome contributions to the [documentation repository](https://github.com/unikraft/docs).
Documentation is written using [MDX](https://mdxjs.com/), since it make it easy to use `JSX` components.

We follow the [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow) for contributions.
Submit new content, corrections and improvements to documentation as pull requests (PRs) that we will then review and approve in the documentation.
You can add entire sections if you think a topic is missing or update existing ones with missing information.

You can also open up [an issue](https://github.com/unikraft/docs/issues) to signal a problem or a missing part in the documentation that you request someone to fix.

We use English as the documentation language.
Please proofread your text before submitting a pull request.

Write each sentence on a new line, since it make the content easy to review and modify.

> This way, changing one sentence only affects one line in the source code.

Reviewer: ```suggestion
This way, changing one sentence only affects one line in the source code, and it is easier to follow the content and to reviews.
```
